# BÀI 1: LÀM QUEN VỚI PYTORCH
Họ tên: Vũ Quốc Anh  MSSV: 3124411016

## Bài 1. Tạo Tensor cơ bản

In [1]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)

print(a)
print("Shape:", a.shape)
print("Ndim:", a.ndim)
print("Dtype:", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Shape: torch.Size([2, 3])
Ndim: 2
Dtype: torch.int32


**Nhận xét:** Tensor `a` có kích thước `torch.Size([2, 3])` (2 hàng, 3 cột), số chiều `ndim = 2`, kiểu dữ liệu `torch.int32` vì được khai báo tường minh bằng `dtype=torch.int32`.

## Bài 2. Tạo các Tensor thường dùng

In [2]:
import torch
torch.manual_seed(0)

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print(a)
print(b)
print(c)
print(d)
print(e)

f = torch.full((5, 5), 10)
print("\nTensor 5x5 gia tri 10:")
print(f)

tensor([[0.4963, 0.7682, 0.0885, 0.1320],
        [0.3074, 0.6341, 0.4901, 0.8964],
        [0.4556, 0.6323, 0.3489, 0.4017]])
tensor([[ 0.4033,  0.8380, -0.7193, -0.4033],
        [-0.5966,  0.1820, -0.8567,  1.1006],
        [-1.0712,  0.1227, -0.5663,  0.3731]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

Tensor 5x5 gia tri 10:
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Nhận xét:**
- `torch.rand()` sinh số ngẫu nhiên phân phối **đều (uniform)** trong khoảng [0, 1) nên các giá trị của `a` luôn dương, nằm trong [0,1).
- `torch.randn()` sinh số ngẫu nhiên phân phối **chuẩn (normal)** với trung bình 0, độ lệch chuẩn 1, nên các giá trị của `b` có thể âm hoặc dương và trải rộng hơn.
- Tensor 5x5 toàn giá trị 10 được tạo bằng `torch.full((5,5), 10)`.

## Bài 3. Kiểm tra thông tin Tensor

In [3]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


**Giải thích `torch.Size([2, 3, 4])`:** Tensor có 3 chiều (ndim = 3). Chiều thứ nhất có kích thước 2 (2 "khối"), chiều thứ hai có kích thước 3 (3 hàng trong mỗi khối), chiều thứ ba có kích thước 4 (4 cột/phần tử trong mỗi hàng). Tổng số phần tử = 2 × 3 × 4 = 24.

## Bài 4. Slicing và Indexing

In [4]:
import torch
torch.manual_seed(1)

a = torch.randn(3, 4, 5)

print("Original tensor:")
print(a)

print("\nTensor a[1]:")
print(a[1])
print("Shape a[1]:", a[1].shape)

print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])
print("Shape a[1:, 2:4]:", a[1:, 2:4].shape)

print("\nHai phan tu cuoi tren chieu thu 3 (a[..., -2:]):")
print(a[..., -2:])
print("Shape:", a[..., -2:].shape)

Original tensor:
tensor([[[-1.5256, -0.7502, -0.6540, -1.6095, -0.1002],
         [-0.6092, -0.9798, -1.6091, -0.7121,  0.3037],
         [-0.7773, -0.2515, -0.2223,  1.6871,  0.2284],
         [ 0.4676, -0.6970, -1.1608,  0.6995,  0.1991]],

        [[ 0.8657,  0.2444, -0.6629,  0.8073,  1.1017],
         [-0.1759, -2.2456, -1.4465,  0.0612, -0.6177],
         [-0.7981, -0.1316,  1.8793, -0.0721,  0.1578],
         [-0.7735,  0.1991,  0.0457,  0.1530, -0.4757]],

        [[-0.1110,  0.2927, -0.1578, -0.0288,  0.4533],
         [ 1.1422,  0.2486, -1.7754, -0.0255, -1.0233],
         [-0.5962, -1.0055,  0.4285,  1.4761, -1.7869],
         [ 1.6103, -0.7040, -0.1853, -0.9962, -0.8313]]])

Tensor a[1]:
tensor([[ 0.8657,  0.2444, -0.6629,  0.8073,  1.1017],
        [-0.1759, -2.2456, -1.4465,  0.0612, -0.6177],
        [-0.7981, -0.1316,  1.8793, -0.0721,  0.1578],
        [-0.7735,  0.1991,  0.0457,  0.1530, -0.4757]])
Shape a[1]: torch.Size([4, 5])

Tensor a[1:, 2:4]:
tensor([[[-0.7981, 

**Nhận xét:**
- `a[1]` lấy khối thứ 2 (index 1) trên chiều đầu tiên, kết quả có shape `[4, 5]` (giảm 1 chiều so với `a`).
- `a[1:, 2:4]` lấy các khối từ index 1 trở đi (chiều 0) và các hàng index 2, 3 (chiều 1), giữ nguyên chiều 2 → shape `[2, 2, 5]`.
- Để lấy 2 phần tử cuối trên chiều thứ ba, dùng slicing `a[..., -2:]`, kết quả có shape `[3, 4, 2]`.

## Bài 5. Reshape và Flatten

In [5]:
import torch
torch.manual_seed(2)

a = torch.randn(3, 4)

print("Original:")
print(a)

print("\nFlatten:")
print(a.ravel())

print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

print("\nSo phan tu truoc reshape:", a.numel())
print("So phan tu sau reshape:", a.reshape(3, 2, 2).numel())

print("\nReshape to 2 x 6:")
print(a.reshape(2, 6))

Original:
tensor([[ 0.3923, -0.2236, -0.3195, -1.2050],
        [ 1.0445, -0.6332,  0.5731,  0.5409],
        [-0.3919, -1.0427,  1.3186,  0.7476]])

Flatten:
tensor([ 0.3923, -0.2236, -0.3195, -1.2050,  1.0445, -0.6332,  0.5731,  0.5409,
        -0.3919, -1.0427,  1.3186,  0.7476])

Reshape to 3 x 2 x 2:
tensor([[[ 0.3923, -0.2236],
         [-0.3195, -1.2050]],

        [[ 1.0445, -0.6332],
         [ 0.5731,  0.5409]],

        [[-0.3919, -1.0427],
         [ 1.3186,  0.7476]]])

So phan tu truoc reshape: 12
So phan tu sau reshape: 12

Reshape to 2 x 6:
tensor([[ 0.3923, -0.2236, -0.3195, -1.2050,  1.0445, -0.6332],
        [ 0.5731,  0.5409, -0.3919, -1.0427,  1.3186,  0.7476]])


**Nhận xét:** Reshape không làm thay đổi tổng số phần tử (đều bằng 12 = 3×4 = 3×2×2 = 2×6), chỉ thay đổi cách sắp xếp các phần tử theo hình dạng mới; dữ liệu được sắp xếp lại theo thứ tự hàng (row-major).

## Bài 6. Phép toán và nhân ma trận

In [6]:
import torch
torch.manual_seed(3)

a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:")
print(a)
print("b:")
print(b)

print("\na + b:")
print(a + b)

print("\na / b:")
print(a / b)

print("\na^2:")
print(a ** 2)

print("\nMatrix multiplication:")
print(a @ b.T)
print("Shape a @ b.T:", (a @ b.T).shape)

print("\na * b (element-wise):")
print(a * b)

a:
tensor([[ 0.8033,  0.1748,  0.0890],
        [-0.6137,  0.0462, -1.3683]])
b:
tensor([[ 0.3375,  1.0111, -1.4352],
        [ 0.9774,  0.5220,  1.2379]])

a + b:
tensor([[ 1.1408,  1.1859, -1.3462],
        [ 0.3637,  0.5682, -0.1303]])

a / b:
tensor([[ 2.3801,  0.1729, -0.0620],
        [-0.6279,  0.0885, -1.1053]])

a^2:
tensor([[0.6453, 0.0306, 0.0079],
        [0.3766, 0.0021, 1.8721]])

Matrix multiplication:
tensor([[ 0.3202,  0.9866],
        [ 1.8033, -2.2695]])
Shape a @ b.T: torch.Size([2, 2])

a * b (element-wise):
tensor([[ 0.2711,  0.1768, -0.1277],
        [-0.5999,  0.0241, -1.6938]])


**Nhận xét:**
- `a * b` là phép **nhân từng phần tử tương ứng** (element-wise), yêu cầu `a` và `b` cùng shape `[2, 3]`, kết quả cũng có shape `[2, 3]`.
- `a @ b.T` là phép **nhân ma trận** thực sự (dot product theo quy tắc đại số tuyến tính): `a` có shape `[2,3]`, `b.T` có shape `[3,2]`, nên kết quả có shape `[2, 2]`.

## Bài 7. Các hàm thống kê trên Tensor

In [7]:
import torch
torch.manual_seed(4)

a = torch.randn(3, 4)
print(a)

print("\nMean (dim=0):")
print(torch.mean(a, dim=0))

print("\nStandard deviation (dim=0):")
print(torch.std(a, dim=0))

print("\nCumulative sum (dim=0):")
print(torch.cumsum(a, dim=0))

print("\n--- dim=1 ---")
print("\nMean (dim=1):")
print(torch.mean(a, dim=1))

print("\nStandard deviation (dim=1):")
print(torch.std(a, dim=1))

print("\nCumulative sum (dim=1):")
print(torch.cumsum(a, dim=1))

tensor([[-1.6053,  0.2325,  2.2399,  0.8473],
        [ 1.2006, -0.4016, -1.4260,  0.9039],
        [ 0.8557,  0.6889,  0.8850,  1.7706]])

Mean (dim=0):
tensor([0.1504, 0.1733, 0.5663, 1.1740])

Standard deviation (dim=0):
tensor([1.5302, 0.5476, 1.8536, 0.5175])

Cumulative sum (dim=0):
tensor([[-1.6053,  0.2325,  2.2399,  0.8473],
        [-0.4046, -0.1691,  0.8139,  1.7512],
        [ 0.4511,  0.5198,  1.6988,  3.5219]])

--- dim=1 ---

Mean (dim=1):
tensor([0.4286, 0.0693, 1.0501])

Standard deviation (dim=1):
tensor([1.5949, 1.2158, 0.4881])

Cumulative sum (dim=1):
tensor([[-1.6053, -1.3728,  0.8671,  1.7144],
        [ 1.2006,  0.7991, -0.6269,  0.2770],
        [ 0.8557,  1.5446,  2.4296,  4.2002]])


**Nhận xét:** `dim=0` tính toán **dọc theo các hàng** (theo từng cột) → kết quả có 4 giá trị (một cho mỗi cột), trong khi `dim=1` tính toán **dọc theo các cột** (theo từng hàng) → kết quả có 3 giá trị (một cho mỗi hàng). Nói cách khác, `dim` chỉ định chiều bị "thu gọn/rút gọn" (collapse) khi tính toán.

## Bài 8. Autograd cơ bản

In [8]:
import torch

x = torch.tensor(
    3.6,
    requires_grad=True
)

y = x * x

y.backward()

print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

print("\n--- Doi ham thanh y = 3x^2 + 2x + 1 ---")
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2**2 + 2 * x2 + 1
y2.backward()
print("x2 =", x2)
print("y2 =", y2)
print("x2.grad =", x2.grad)
print("Dao ham thu cong dy2/dx = 6x + 2 tai x=3.6:", 6*3.6 + 2)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)

--- Doi ham thanh y = 3x^2 + 2x + 1 ---
x2 = tensor(3.6000, requires_grad=True)
y2 = tensor(47.0800, grad_fn=<AddBackward0>)
x2.grad = tensor(23.6000)
Dao ham thu cong dy2/dx = 6x + 2 tai x=3.6: 23.6


**Nhận xét:**
- Với y = x², đạo hàm dy/dx = 2x. Tại x = 3.6: 2×3.6 = 7.2, khớp với `x.grad = 7.2000` mà autograd tính được.
- Với y = 3x² + 2x + 1, đạo hàm dy/dx = 6x + 2. Tại x = 3.6: 6×3.6 + 2 = 23.6, khớp chính xác với `x2.grad = 23.6000`. Điều này xác nhận autograd tính đạo hàm tự động chính xác bằng quy tắc chuỗi (chain rule).

## Bài 9. Gradient Descent tìm hệ số đa thức

In [9]:
import numpy as np
import torch
np.random.seed(0)
torch.manual_seed(0)

polynomial = np.poly1d([1, 2, 3])
N = 20

X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients:")
print(w)

Initial coefficients:
tensor([[ 1.5410],
        [-0.2934],
        [-2.1788]], requires_grad=True)


Learned coefficients:
tensor([[1.0017],
        [2.0058],
        [2.8337]], requires_grad=True)


**Nhận xét:**
- Ba hệ số học được `[1.0017, 2.0058, 2.8337]` rất gần với bộ hệ số thật `[1, 2, 3]` của đa thức y = x² + 2x + 3, sai lệch nhỏ chủ yếu do số epoch huấn luyện (1000) chưa hội tụ tuyệt đối và learning rate được chọn cố định.
- **`optimizer.zero_grad()`**: đặt lại gradient tích lũy về 0 trước mỗi lượt lan truyền ngược, vì PyTorch mặc định **cộng dồn (accumulate)** gradient qua các lần gọi `backward()`.
- **`mse.backward()`**: thực hiện lan truyền ngược (backpropagation), tính gradient của hàm mất mát `mse` theo từng tham số (ở đây là `w`) và lưu vào thuộc tính `.grad`.
- **`optimizer.step()`**: dùng gradient vừa tính được để cập nhật giá trị tham số `w` theo thuật toán tối ưu (ở đây là NAdam), nhằm giảm dần giá trị hàm mất mát.

## Bài 10. Giải hệ phương trình bằng Autograd

In [10]:
import random
import torch
random.seed(0)
torch.manual_seed(0)

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2

    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

print("\nKiem tra:")
print("A+B =", (A+B).item(), " (ky vong 9)")
print("C-D =", (C-D).item(), " (ky vong 1)")
print("A+C =", (A+C).item(), " (ky vong 8)")
print("B-D =", (B-D).item(), " (ky vong 2)")

A = tensor(4.5494, requires_grad=True)
B = tensor(4.4506, requires_grad=True)
C = tensor(3.4506, requires_grad=True)
D = tensor(2.4506, requires_grad=True)

Kiem tra:
A+B = 8.99995231628418  (ky vong 9)
C-D = 1.0000104904174805  (ky vong 1)
A+C = 7.999969005584717  (ky vong 8)
B-D = 1.9999933242797852  (ky vong 2)


**Nhận xét:**
- Bốn giá trị kiểm tra `A+B ≈ 9`, `C-D ≈ 1`, `A+C ≈ 8`, `B-D ≈ 2` đều xấp xỉ đúng với yêu cầu đề bài (sai số rất nhỏ, cỡ 10⁻⁴–10⁻⁵), cho thấy autograd + optimizer đã tối ưu tốt tổng bình phương sai số về gần 0.
- Hệ có **4 ẩn (A, B, C, D)** nhưng chỉ có **4 phương trình**, tuy nhiên các phương trình này **không độc lập tuyến tính** hoàn toàn theo nghĩa xác định duy nhất 4 giá trị — thực chất hệ có vô số nghiệm vì có thể cộng cùng một hằng số k vào cả 4 biến A, B, C, D mà vẫn thỏa mãn hệ (do các phương trình chỉ ràng buộc *hiệu* hoặc *tổng* giữa các cặp biến theo dạng sai khác không làm thay đổi kết quả khi dịch chuyển đồng thời). Vì vậy nghiệm tìm được phụ thuộc vào điểm khởi tạo ngẫu nhiên ban đầu.

## Bài 11. Bài tập áp dụng

In [11]:
import numpy as np
import torch
torch.manual_seed(5)

print("===== Yeu cau 1: Reshape Tensor X 24 phan tu =====")
X = torch.arange(24)
print("X ban dau:", X)
print("Reshape 4x6:")
print(X.reshape(4, 6))
print("Reshape 2x12:")
print(X.reshape(2, 12))
print("Reshape 2x3x4:")
print(X.reshape(2, 3, 4))

===== Yeu cau 1: Reshape Tensor X 24 phan tu =====
X ban dau: tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
Reshape 4x6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2x12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2x3x4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


In [12]:
print("===== Yeu cau 2: A, B 3x4 =====")
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("A:")
print(A)
print("B:")
print(B)
print("A + B:")
print(A + B)
print("A * B:")
print(A * B)
print("A / B:")
print(A / B)
print("Trung binh theo cot (dim=0):")
print(torch.mean(A, dim=0))
print("Do lech chuan theo cot (dim=0):")
print(torch.std(A, dim=0))

===== Yeu cau 2: A, B 3x4 =====
A:
tensor([[-0.4868, -0.6038, -0.5581,  0.6675],
        [-0.1974,  1.9428, -1.4017, -0.7626],
        [ 0.6312, -0.8991, -0.5578,  0.6907]])
B:
tensor([[ 0.2225, -0.6662,  0.6846,  0.5740],
        [-0.5829,  0.7679,  0.0571, -1.1894],
        [-0.5659, -0.8327,  0.9014,  0.2116]])
A + B:
tensor([[-0.2643, -1.2700,  0.1265,  1.2415],
        [-0.7803,  2.7107, -1.3446, -1.9519],
        [ 0.0653, -1.7318,  0.3437,  0.9023]])
A * B:
tensor([[-0.1083,  0.4023, -0.3821,  0.3832],
        [ 0.1151,  1.4919, -0.0800,  0.9070],
        [-0.3572,  0.7487, -0.5028,  0.1461]])
A / B:
tensor([[ -2.1882,   0.9063,  -0.8152,   1.1629],
        [  0.3387,   2.5299, -24.5514,   0.6411],
        [ -1.1154,   1.0798,  -0.6188,   3.2647]])
Trung binh theo cot (dim=0):
tensor([-0.0177,  0.1466, -0.8392,  0.1986])
Do lech chuan theo cot (dim=0):
tensor([0.5803, 1.5625, 0.4871, 0.8324])


In [13]:
print("===== Yeu cau 3: Dao ham y = 2x^2 + 5x + 3 tai x=2 =====")
x = torch.tensor(2.0, requires_grad=True)
y = 2 * x**2 + 5 * x + 3
y.backward()
print("y =", y.item())
print("dy/dx (autograd) =", x.grad.item())
print("dy/dx (thu cong) = 4x + 5 tai x=2 =", 4*2 + 5)

===== Yeu cau 3: Dao ham y = 2x^2 + 5x + 3 tai x=2 =====
y = 21.0
dy/dx (autograd) = 13.0
dy/dx (thu cong) = 4x + 5 tai x=2 = 13


In [14]:
print("===== Yeu cau 4: Gradient Descent voi y = 2x^2 - 3x + 5 =====")
np.random.seed(1)
polynomial = np.poly1d([2, -3, 5])
N = 20
Xd = np.random.randn(N, 1) * 5
Yd = polynomial(Xd)

XX = np.hstack([Xd * Xd, Xd, np.ones_like(Xd)])

w = torch.randn(3, 1, requires_grad=True)
xt = torch.tensor(XX, dtype=torch.float32)
yt = torch.tensor(Yd, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)
print("He so ban dau:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = xt @ w
    mse = torch.mean(torch.square(yt - y_pred))
    mse.backward()
    optimizer.step()

print("He so hoc duoc (ky vong [2, -3, 5]):")
print(w)

===== Yeu cau 4: Gradient Descent voi y = 2x^2 - 3x + 5 =====
He so ban dau:
tensor([[ 0.4479],
        [-0.6088],
        [ 0.2389]], requires_grad=True)


He so hoc duoc (ky vong [2, -3, 5]):
tensor([[ 2.0336],
        [-2.9710],
        [ 2.6074]], requires_grad=True)


**Nhận xét chung Bài 11:**
- Reshape Tensor 24 phần tử thành `4×6`, `2×12`, `2×3×4` đều giữ nguyên thứ tự phần tử và tổng số phần tử (24).
- Các phép toán `A+B`, `A*B`, `A/B` trên hai Tensor `3×4` đều là phép toán **theo từng phần tử**, kết quả cùng shape `[3,4]`.
- Đạo hàm autograd của y = 2x² + 5x + 3 tại x = 2 cho kết quả **13.0**, khớp hoàn toàn với tính tay dy/dx = 4x + 5 = 4×2+5 = 13.
- Huấn luyện lại với đa thức y = 2x² − 3x + 5 (1000 epoch), hệ số học được xấp xỉ **[2.03, −2.97, 2.61]** — hai hệ số đầu (bậc 2 và bậc 1) hội tụ khá tốt về [2, −3], còn hệ số tự do (bằng 5) hội tụ chậm hơn nên còn lệch; nếu tăng số epoch (thử 3000 epoch) hệ số tự do tiệm cận gần hơn về 5 (khoảng 4.07), cho thấy việc chọn số epoch và learning rate ảnh hưởng trực tiếp đến độ chính xác của kết quả hồi quy.